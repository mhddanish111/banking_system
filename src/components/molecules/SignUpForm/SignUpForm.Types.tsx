export enum IncomeEnum {
  belowOne = 'below 1 Lac',
  oneToFive = '1 Lac to 5 Lac',
  fiveToTen = '5 Lac to 10 Lac',
  moreThanTen = 'more than 10 Lac',
}

export enum GovtIdEnum {
  aadharCard = 'Aadhar Card',
  panCard = 'Pan Card',
  passport = 'Passport',
  electionIdCard = 'Election Commission ID Card',
}

export type FormData = {
  firstName: string;
  lastName: string;
  email: string;
  mobile: number;
  address: string;
  city: string;
  state: string;
  postalCode: number;
  occupation: string;
  income: IncomeEnum;
  password: string;
  confirmPassword: string;
  govtId: GovtIdEnum;
  govtIdNumber: string;
  aadhaarNumber: string;
  panNumber: string;
};

export type CaptchaProps = {
  valid: boolean;
  showErrorMsg: boolean;
};

export type SignUpFormProps = {
  SignUpHandler: (e: any, data: FormData) => void;
};
