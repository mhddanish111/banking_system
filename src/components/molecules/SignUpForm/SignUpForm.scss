@import "../../../styles/spacing.scss";
@import '../../../styles/boarder.scss';
@import "../../../styles/colors.scss";
.sign-up-form {
  width: 100%;
  display: flex;
  justify-content: center;
  padding: 20px;

  button {
    width: 100%;
  }

  .form-input {
    display: flex;
    flex-direction: column;
    margin-bottom: $space-m;
  }
  .form-action {
    margin-top: $space-m
  }

  .sign-in-link {
    margin-top: $space-xs;
    margin-bottom: $space-m;
  }

  .error-message {
    color: $color-red-50;
  }
}
