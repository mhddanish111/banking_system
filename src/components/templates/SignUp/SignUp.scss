@import '../../../styles/boarder.scss';
@import '../../../styles/spacing.scss';
@import '../../../styles/mixins.scss';
.sign-up {
    .main-container {
        display: flex;
        justify-content: center;
    }
}
.sign-up-container {
    border: $border-medium;
    border-radius: $border-radius;
    box-shadow: $box-shadow-focus;
    min-height: 500px;
    width: 60%;
    margin-top: $space-xxl;
    margin-bottom: $space-xxl;
    @include mobile {
        width: 85%;
    }
    @include tablet {
        width: 80%;
    }
    .error-message-container{
        margin: 0 ($space-l - $space-xxs);
    }
}

.sign-up-heading {
    display: flex;
    justify-content:flex-start;
    margin: $space-xxs;
    padding: $space-xxs;
}