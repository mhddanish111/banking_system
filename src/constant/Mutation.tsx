import { gql } from '@apollo/client';

export const CREATE_USER = gql`
  mutation CreateUser($userInput: UserInput) {
    createUser(userInput: $userInput) {
      firstName
      lastName
      message
    }
  }
`;

export const CHANGE_PASSWORD = gql`
  mutation ChangePassword($changePasswordInput: ChangePasswordInput) {
    changePassword(changePasswordInput: $changePasswordInput) {
      message
      success
      data {
        firstName
      }
    }
  }
`;

export const OPEN_ACCOUNT = gql`
  mutation CreateAccount($accountInput: AccountInput) {
    createAccount(accountInput: $accountInput) {
      _id
      accountNumber
      amount
      accountType
      categoryName
      limit
      minAmount
      maxAmount
      accountTypeId
      categoryId
      userId
      message
      type
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
    }
  }
`;

export const DEPOSIT = gql`
  mutation Deposit($depositInput: DepositInput) {
    deposit(depositInput: $depositInput) {
      transactionResult {
        _id
      }
      accounts {
        _id
        type
        accountType
        categoryName
        accountNumber
        amount
        minAmount
        maxAmount
        limit
        charge
        debitCardNumber
        debitCardCSVNumber
        expiryYear
        expiryMonth
        users {
          _id
          firstName
          lastName
        }
      }
    }
  }
`;

export const WITHDRAWAL = gql`
  mutation Withdrawal($withdrawalInput: WithdrawalInput) {
    withdrawal(withdrawalInput: $withdrawalInput) {
      transactionResult {
        _id
      }
      accounts {
        _id
        type
        accountType
        categoryName
        accountNumber
        amount
        minAmount
        maxAmount
        limit
        charge
        debitCardNumber
        debitCardCSVNumber
        expiryYear
        expiryMonth
        users {
          _id
          firstName
          lastName
        }
      }
    }
  }
`;
