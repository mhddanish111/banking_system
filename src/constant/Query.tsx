import { gql } from '@apollo/client';

export const LOGIN_USER = gql`
  query Login($logInInput: LogInInput) {
    login(logInInput: $logInInput) {
      firstName
      lastName
      userId
      token
      isValid
      message
      userType
    }
  }
`;

export const USER_DETAILS = gql`
  query UserDetails {
    userDetails {
      _id
      firstName
      lastName
      email
      mobile
      occupation
      income
      aadhaarNumber
      panNumber
      govtId
      govtIdNumber
      address
      city
      state
      postalCode
      isActive
      message
    }
  }
`;

export const GET_ACCOUNT_TYPES = gql`
  query AccountTypes {
    accountType {
      _id
      name
      type
      isActive
      category {
        _id
        name
        benefit
        isActive
        minAmount
        maxAmount
        limit
        charge
      }
    }
  }
`;

export const GET_ACCOUNTS = gql`
  query Accounts {
    accounts {
      _id
      type
      accountType
      categoryName
      accountNumber
      amount
      minAmount
      maxAmount
      limit
      charge
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
      users {
        _id
        firstName
        lastName
      }
    }
  }
`;

export const GET_ACCOUNT_WITH_TYPE = gql`
  query AccountWithType($accountWithTypeInput: AccountWithTypeInput) {
    accountWithType(accountWithTypeInput: $accountWithTypeInput) {
      _id
      type
      accountType
      categoryName
      accountNumber
      amount
      minAmount
      maxAmount
      limit
      charge
      debitCardNumber
      debitCardCSVNumber
      expiryYear
      expiryMonth
      users {
        _id
        firstName
        lastName
        aadhaarNumber
        panNumber
      }
    }
  }
`;

export const GET_TRANSACTION = gql`
  query Transaction($transactionInput: TransactionInput) {
    transaction(transactionInput: $transactionInput) {
      count {
        offset
        limit
        totalCount
      }
      result {
        _id
        transactionId
        accountNumber
        amount
        withdrawalAmount
        depositAmount
        transactionDate
        accountType
        categoryName
        transactionType
      }
    }
  }
`;
