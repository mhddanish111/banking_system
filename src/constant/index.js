export const USER_KEY = 'user_key';

export const AccountTypeList = [
  { label: 'Accounts', value: 'Accounts' },
  { label: 'Cards', value: 'Cards' },
];
