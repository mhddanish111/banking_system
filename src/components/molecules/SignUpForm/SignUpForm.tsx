import React, { useState } from 'react';
import { isEmpty } from 'lodash';
import { Link } from 'react-router-dom';
import { useForm } from 'react-hook-form';
import Button from '../../atoms/Button';
import Captcha from '../../atoms/Captcha';
import { FormData, CaptchaProps, SignUpFormProps } from './SignUpForm.Types';
import './SignUpForm.scss';

const SignUpForm: React.FC<SignUpFormProps> = ({ SignUpHandler }) => {
  const [password, setPassword] = useState<String>();
  const [govId, setGovId] = useState<String>('');
  const [reCaptchaStatus, setReCaptchaStatus] = useState<CaptchaProps>({
    valid: false,
    showErrorMsg: false,
  });
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormData>({ mode: 'onChange' });
  const onSubmit = (data: FormData, e: any) => {
    if (!reCaptchaStatus.valid) {
      return setReCaptchaStatus({ valid: false, showErrorMsg: true });
    }
    if (data.password !== data.confirmPassword) {
      return;
    }
    SignUpHandler(e, data);
  };
  const getErrorMessage = (message: String | undefined) => {
    return message && <p className="error-message">{message}</p>;
  };
  return (
    <form
      name="sign-up-form"
      className="row sign-up-form"
      onSubmit={handleSubmit(onSubmit)}
      autoComplete="off"
      aria-label="sign in form"
    >
      <div className="col-sm-6 col-xs-12">
        <label htmlFor="firstName" aria-label="first name">
          First Name
        </label>
        <input
          type="text"
          id="firstName"
          autoComplete="off"
          {...register('firstName', {
            required: 'First name is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 5 !',
            },
          })}
        />
        {getErrorMessage(errors?.firstName?.message)}
      </div>
      <div className="col-sm-6 col-xs-12">
        <label htmlFor="lastName" aria-label="last name">
          Last Name
        </label>
        <input
          type="text"
          id="lastName"
          autoComplete="off"
          {...register('lastName', {
            required: 'Last name is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 50 !',
            },
          })}
        />
        {getErrorMessage(errors?.lastName?.message)}
      </div>
      <div className="col-sm-6 col-xs-12">
        <label htmlFor="email" aria-label="email">
          Email
        </label>
        <input
          type="email"
          id="email"
          autoComplete="off"
          {...register('email', {
            required: 'Email is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 50 !',
            },
          })}
        />
        {getErrorMessage(errors?.email?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="mobile" aria-label="mobile number">
          Mobile
        </label>
        <input
          type="number"
          id="mobile"
          autoComplete="off"
          {...register('mobile', {
            required: 'Mobile is required',
            maxLength: {
              value: 10,
              message: 'You exceeded max length of 10 !',
            },
          })}
        />
        {getErrorMessage(errors?.mobile?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="occupation" aria-label="occupation">
          Occupation
        </label>
        <input
          type="text"
          id="occupation"
          autoComplete="off"
          {...register('occupation', {
            required: 'Occupation is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 50 !',
            },
          })}
        />
        {getErrorMessage(errors?.occupation?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="income" aria-label="income">
          Income
        </label>
        <select {...register('income', { required: 'Income is required' })}>
          <option value="">Select</option>
          <option value="belowOne">below 1 Lac</option>
          <option value="oneToFive">1 Lac to 5 Lac</option>
          <option value="fiveToTen">5 Lac to 10 Lac</option>
          <option value="moreThanTen">more than 10 Lac</option>
        </select>
        {getErrorMessage(errors?.income?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="aadhaarNumber" aria-label="aadhaar number">
          Aadhaar Number
        </label>
        <input
          type="text"
          id="aadhaarNumber"
          autoComplete="off"
          {...register('aadhaarNumber', {
            required: 'aadhaar number is required',
            // pattern: {
            //   value: /^[2-9]{1}[0-9]{3}\s[0-9]{4}\s[0-9]{4}$/,
            //   message: "Aadhaar number is not valid",
            // },
            maxLength: {
              value: 15,
              message: 'You exceeded max length!',
            },
          })}
        />
        {getErrorMessage(errors?.aadhaarNumber?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="panNumber" aria-label="pan card number">
          Pan card Number
        </label>
        <input
          type="text"
          id="panNumber"
          autoComplete="off"
          {...register('panNumber', {
            required: 'pan card number is required',
            // pattern: {
            //   value: /^[A-Z]{5}[0-9]{4}[A-Z]{1}$/,
            //   message: "Pan card number is not valid",
            // },
            maxLength: {
              value: 10,
              message: 'You exceeded max length!',
            },
          })}
        />
        {getErrorMessage(errors?.panNumber?.message)}
      </div>

      <div
        className={`${!isEmpty(govId) ? 'col-sm-6' : 'col-sm-12'} col-xs-12`}
      >
        <label htmlFor="govtId" aria-label="government id">
          Govt ID
        </label>
        <select
          {...register('govtId', {
            onChange: (e) => setGovId(e.target.value),
          })}
        >
          <option value="">Select</option>
          <option value="passport">Passport</option>
          <option value="electionIdCard">Election Commission ID Card</option>
        </select>
        {getErrorMessage(errors?.govtId?.message)}
      </div>

      {!isEmpty(govId) && (
        <div className="col-sm-6 col-xs-12">
          <label htmlFor="govtIdNumber" aria-label="government ID number">
            Government ID Number
          </label>
          <input
            type="text"
            id="govtIdNumber"
            autoComplete="off"
            {...register('govtIdNumber', {
              required: !isEmpty(govId)
                ? 'Government ID Number is required'
                : false,
              maxLength: {
                value: 30,
                message: 'You exceeded max length of 30 !',
              },
            })}
          />
          {getErrorMessage(errors?.govtIdNumber?.message)}
        </div>
      )}

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="address" aria-label="address">
          Address
        </label>
        <input
          type="text"
          id="address"
          autoComplete="off"
          {...register('address', {
            required: 'Address is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 50 !',
            },
          })}
        />
        {getErrorMessage(errors?.address?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="city" aria-label="city">
          City
        </label>
        <input
          type="text"
          id="city"
          autoComplete="off"
          {...register('city', {
            required: 'City is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 50 !',
            },
          })}
        />
        {getErrorMessage(errors?.city?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="state" aria-label="state">
          State
        </label>
        <input
          type="text"
          id="state"
          autoComplete="off"
          {...register('state', {
            required: 'State is required',
            maxLength: {
              value: 50,
              message: 'You exceeded max length of 50 !',
            },
          })}
        />
        {getErrorMessage(errors?.state?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="postalCode" aria-label="postal code">
          Postal Code
        </label>
        <input
          type="number"
          id="postalCode"
          autoComplete="off"
          {...register('postalCode', {
            required: 'Postal code is required',
            maxLength: {
              value: 6,
              message: 'You exceeded max length of 6 !',
            },
          })}
        />
        {getErrorMessage(errors?.postalCode?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="password" aria-label="password">
          Password
        </label>
        <input
          type="password"
          id="password"
          autoComplete="off"
          {...register('password', {
            onBlur: (e) => setPassword(e.target.value),
            required: 'Password is required',
            maxLength: {
              value: 15,
              message: 'You exceeded max length of 15 !',
            },
          })}
        />
        {getErrorMessage(errors?.password?.message)}
      </div>

      <div className="col-sm-6 col-xs-12">
        <label htmlFor="confirmPassword" aria-label="confirm password">
          Confirm Password
        </label>
        <input
          type="password"
          id="confirmPassword"
          autoComplete="off"
          {...register('confirmPassword', {
            required: 'Confirm password is required',
            validate: (value) =>
              value === password || 'Password does not match',
            maxLength: {
              value: 15,
              message: 'You exceeded max length of 15 !',
            },
          })}
        />
        {getErrorMessage(errors?.confirmPassword?.message)}
      </div>
      <div className="col-xs-12 form-action">
        <Captcha setReCaptchaStatus={setReCaptchaStatus} />
        {reCaptchaStatus.showErrorMsg &&
          !reCaptchaStatus.valid &&
          getErrorMessage('Captcha is not valid')}
      </div>

      <div className="col-xs-12 form-action">
        <Button type="submit">Sign Up</Button>
      </div>
      <div className="col-xs-12 sign-in-link">
        <Link to="/">Sign In</Link>
      </div>
    </form>
  );
};

export default SignUpForm;
