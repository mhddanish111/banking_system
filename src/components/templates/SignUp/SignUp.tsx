import React, { useState } from 'react';
import { useMutation } from '@apollo/client';
import { Link } from 'react-router-dom';
import Header from '../../organisms/Header';
import Heading from '../../atoms/Heading';
import SignUpForm from '../../molecules/SignUpForm';
import ErrorMessage from '../../atoms/ErrorMessage';
import { FormData } from '../../molecules/SignUpForm/SignUpForm.Types';
import { CREATE_USER } from '../../../constant/Mutation';
import './SignUp.scss';

const SignUp: React.FC<any> = (): any => {
  const [createUserHandler] = useMutation(CREATE_USER);
  const [errorMessage, setErrorMessage] = useState<String>('');
  const [isSignUpSuccess, setSignUpSuccess] = useState<Boolean>(false);
  const SignUpHandler = (e: any, data: FormData) => {
    e.preventDefault();
    createUserHandler({ variables: { userInput: data } })
      .then((res) => {
        const { message } = res?.data?.createUser;
        if (message) {
          setErrorMessage(message);
          window.scroll(0, 0);
        } else {
          setErrorMessage('');
          setSignUpSuccess(true);
        }
      })
      .catch((err) => {
        setErrorMessage(err?.message);
      });
  };
  const closeError = (e: any) => {
    e.preventDefault();
    setErrorMessage('');
  };
  return (
    <div className="sign-up">
      <Header isDefault={false} />
      <main className="main-container">
        <div className="sign-up-container">
          <Heading
            HeadingType="h1"
            className="sign-up-heading"
            ariaLabel="sign up title"
          >
            Create Account
          </Heading>
          {errorMessage && (
            <ErrorMessage
              message={errorMessage}
              className="alert"
              onClick={closeError}
            />
          )}
          {isSignUpSuccess && (
            <ErrorMessage
              message="Welcome to Publicis Sapient Bank, Please login"
              className="success"
              onClick={closeError}
            >
              <Link to="/">Sign In</Link>
            </ErrorMessage>
          )}
          {!isSignUpSuccess && <SignUpForm SignUpHandler={SignUpHandler} />}
        </div>
      </main>
    </div>
  );
};

export default SignUp;
