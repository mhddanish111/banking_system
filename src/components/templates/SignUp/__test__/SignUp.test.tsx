import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import { MemoryRouter, BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import SignUp from '../SignUp';
import { CREATE_USER } from '../../../../constant/Mutation';
import SignUpForm from '../../../molecules/SignUpForm';

describe('<SignIn />', () => {
  const props = {
    selectedAccountNumber: 1673894730214,
    setIsCallTransaction: jest.fn(),
    isCallTransaction: true,
  };
  const data = {
    firstName: 'test',
    lastName: 'test',
    email: 'test@test',
    mobile: 99999999,
    address: 'test',
    city: 'test',
    state: 'test',
    postalCode: 909090,
    occupation: 'test',
    income: 'test',
    password: 'test',
    confirmPassword: 'test',
    govtId: 'test',
    govtIdNumber: 'test',
    aadhaarNumber: 'test',
    panNumber: 'test',
  };
  const mocks = [
    {
      request: {
        query: CREATE_USER,
        variables: {
          userInput: data,
        },
      },
      result: {
        data: {
          createUser: {
            message: 'test',
          },
        },
      },
    },
  ];
  const mockedUsedNavigate = jest.fn();
  jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: () => mockedUsedNavigate,
  }));
  const wrapper = renderer.create(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <SignUp {...props} />
      </BrowserRouter>
    </MockedProvider>,
  ).toJSON;
  render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>
        <SignUp {...props} />
      </BrowserRouter>
    </MockedProvider>,
  );
  it('should render the SignUp Component correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should call SignUpForm Component correctly', async () => {
    const SignUpHandler = jest.fn();
    render(<SignUpForm SignUpHandler={SignUpHandler} />, {
      wrapper: MemoryRouter,
    });
    await userEvent.click(screen.getByRole('button'));
  });
});
