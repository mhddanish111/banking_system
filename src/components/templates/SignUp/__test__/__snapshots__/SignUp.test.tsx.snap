// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn /> should render the SignUp Component correctly 1`] = `[Function]`;
